src/index.js

```diff
@@ -1,390 +1,603 @@
+    } else if (sub[k] === oldSub[k] || isSameAction(sub[k], oldSub[k])) {
     } else {
-      updateElement(
-        element,
-        oldNode.attributes,
-        node.attributes,
-        (isSvg = isSvg || node.nodeName === "svg")
-      )
-
-      var oldKeyed = {}
-      var newKeyed = {}
-      var oldElements = []
-      var oldChildren = oldNode.children
-      var children = node.children
+      cancel(oldSub)
+      return start(sub, dispatch)
+    }
+  }
+  return oldSub
+}
 
-      for (var i = 0; i < oldChildren.length; i++) {
-        oldElements[i] = element.childNodes[i]
+var start = function(sub, dispatch) {
+  return merge(sub, {
+    cancel: sub.effect(sub, dispatch)
+  })
+}
 
-        var oldKey = getKey(oldChildren[i])
-        if (oldKey != null) {
-          oldKeyed[oldKey] = [oldElements[i], oldChildren[i]]
-        }
-      }
+var refresh = function(sub, oldSub, dispatch) {
+  if (isArray(sub) || isArray(oldSub)) {
+    var out = []
+    var subs = isArray(sub) ? sub : [sub]
+    var oldSubs = isArray(oldSub) ? oldSub : [oldSub]
 
-      var i = 0
-      var k = 0
+    for (var i = 0; i < subs.length || i < oldSubs.length; i++) {
+      out.push(refresh(subs[i], oldSubs[i], dispatch))
+    }
 
-      while (k < children.length) {
-        var oldKey = getKey(oldChildren[i])
-        var newKey = getKey((children[k] = resolveNode(children[k])))
+    return out
+  }
 
-        if (newKeyed[oldKey]) {
-          i++
-          continue
-        }
+  return sub
+    ? oldSub
+      ? restart(sub, oldSub, dispatch)
+      : start(sub, dispatch)
+    : oldSub
+      ? cancel(oldSub)
+      : oldSub
+}
 
-        if (newKey != null && newKey === getKey(oldChildren[i + 1])) {
-          if (oldKey == null) {
-            removeElement(element, oldElements[i], oldChildren[i])
-          }
-          i++
-          continue
-        }
+export function app(props) {
+  var state
+  var view = props.view
+  var subs = props.subscriptions
+  var container = props.container
+  var element = container.children[0]
+  var lastNode = element && recycleElement(element)
+  var lastSub = []
+  var updateInProgress = false
+
+  var setState = function(newState) {
+    if (state !== newState) {
+      state = newState
+
+      if (!updateInProgress) {
+        updateInProgress = true
+        defer(render)
+      }
+    }
+  }
 
-        if (newKey == null || isRecycling) {
-          if (oldKey == null) {
-            patch(element, oldElements[i], oldChildren[i], children[k], isSvg)
-            k++
-          }
-          i++
-        } else {
-          var keyedNode = oldKeyed[newKey] || []
+  var dispatch = function(obj, data) {
+    if (obj == null) {
+    } else if (typeof obj === "function") {
+      dispatch(obj(state, data))
+    } else if (isArray(obj)) {
+      if (typeof obj[0] === "function") {
+        dispatch(obj[0](state, obj[1], data))
+      } else {
+        obj[1].effect(obj[1], dispatch, setState(obj[0]))
```

Comment:
Wasn't there talk of using tuples for the effect and props instead of an object container?
```suggestion
         obj[1].[0](obj[1][1], dispatch, setState(obj[0]))
```

Or was that already debated and rejected? :thinking:

Reply:
@okwolf Not rejected at all. Going to use tuples as originally suggested by @sergey-shpak.